ekm/remote_key_manager.go

```diff
@@ -0,0 +1,455 @@
+	}
+
+	if statuses[0] != ssvsignerclient.StatusImported {
+		return fmt.Errorf("unexpected status %s", statuses[0])
+	}
+
+	if err := km.BumpSlashingProtection(sharePubKey); err != nil {
+		return fmt.Errorf("could not bump slashing protection: %w", err)
+	}
+
+	return nil
+}
+
+func (km *RemoteKeyManager) RemoveShare(pubKey []byte) error {
+	statuses, err := km.client.RemoveValidators(context.Background(), pubKey) // TODO: use context
+	if err != nil {
+		return fmt.Errorf("remove validator: %w", err)
+	}
+
+	if statuses[0] != ssvsignerclient.StatusDeleted {
+		return fmt.Errorf("received status %s", statuses[0])
+	}
+
+	if err := km.RemoveHighestAttestation(pubKey); err != nil {
+		return fmt.Errorf("could not remove highest attestation: %w", err)
+	}
+	if err := km.RemoveHighestProposal(pubKey); err != nil {
+		return fmt.Errorf("could not remove highest proposal: %w", err)
+	}
+
+	return nil
+}
+
+func (km *RemoteKeyManager) retryFunc(f func(arg any) (any, error), arg any, funcName string) (any, error) {
+	if km.retryCount < 2 {
+		return f(arg)
+	}
+
+	var multiErr error
+	for i := 1; i <= km.retryCount; i++ {
+		v, err := f(arg)
+		if err == nil {
+			return v, nil
+		}
+		var shareDecryptionError ssvsignerclient.ShareDecryptionError
+		if errors.As(err, &shareDecryptionError) {
+			return nil, ShareDecryptionError(err)
+		}
+		multiErr = errors.Join(multiErr, err)
+		km.logger.Warn("call failed", zap.Error(err), zap.Int("attempt", i), zap.String("func", funcName))
+	}
+
+	return nil, fmt.Errorf("no successful result after %d attempts: %w", km.retryCount, multiErr)
+}
+
+func (km *RemoteKeyManager) SignBeaconObject(
+	obj ssz.HashRoot,
+	domain phase0.Domain,
+	sharePubkey []byte,
+	signatureDomain phase0.DomainType,
+) (spectypes.Signature, [32]byte, error) {
+	forkInfo, err := km.getForkInfo(context.Background()) // TODO: consider passing context to SignBeaconObject
+	if err != nil {
+		return spectypes.Signature{}, [32]byte{}, fmt.Errorf("get fork info: %w", err)
+	}
+
+	req := web3signer.SignRequest{
+		ForkInfo: forkInfo,
+	}
+
+	switch signatureDomain {
+	case spectypes.DomainAttester:
+		data, ok := obj.(*phase0.AttestationData)
+		if !ok {
+			return nil, [32]byte{}, errors.New("could not cast obj to AttestationData")
+		}
+
+		if err := km.IsAttestationSlashable(sharePubkey, data); err != nil {
+			return nil, [32]byte{}, err
+		}
+
+		if err := km.UpdateHighestAttestation(sharePubkey, data); err != nil {
+			return nil, [32]byte{}, err
+		}
+
+		req.Type = web3signer.Attestation
+		req.Attestation = data
+
+	case spectypes.DomainProposer:
+		switch v := obj.(type) {
+		case *capella.BeaconBlock, *deneb.BeaconBlock:
+			return nil, [32]byte{}, fmt.Errorf("web3signer supports only blinded blocks since bellatrix") // https://github.com/Consensys/web3signer/blob/85ed009955d4a5bbccba5d5248226093987e7f6f/core/src/main/java/tech/pegasys/web3signer/core/service/http/handlers/signing/eth2/BlockRequest.java#L29
```

Comment:
does that mean we aren't supporting signing local blocks with remote-signer? this sounds odd to me.

Reply:
Yeah, actually I don't see even `altair` and `bellatrix` support, the type they are using matches `phase0.BeaconBlock`: https://github.com/Consensys/teku/blob/8479214730b6ab85c6e920998d9010f38461cd19/data/serializer/src/main/java/tech/pegasys/teku/api/schema/BeaconBlockBody.java#L46

Reply:
https://github.com/ssvlabs/ssv/pull/2028#discussion_r2010666012